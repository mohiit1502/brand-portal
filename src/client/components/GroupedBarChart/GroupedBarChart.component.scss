.c-GroupedBarChart {
  //path {
  //  stroke: steelblue;
  //  stroke-width: 2;
  //  fill: none;
  //}
  //.axis path,
  //.axis line {
  //  fill: none;
  //  stroke: grey;
  //  stroke-width: 1;
  //  shape-rendering: crispEdges;
  //}
}

.d3-tip{
  z-index: 10;
}

.d3-tip:after {
  box-sizing: border-box;
  display: inline;
  font-size: 10px;
  width: 100%;
  line-height: 1;
  color: rgba(0, 0, 0, 0.8);
  content: "\25BC";
  position: absolute;
  text-align: center;
}
