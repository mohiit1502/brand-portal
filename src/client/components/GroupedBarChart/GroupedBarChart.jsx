import React, {memo, useEffect, useRef} from "react";
import { renderToString } from 'react-dom/server'
import PropTypes from "prop-types";
import {axisBottom, axisLeft, format, max, scaleBand, scaleLinear, select} from "d3";
import d3Tip from "d3-tip"
import useResizeObserver from "../../hooks/useResizeObserver";
import "./GroupedBarChart.component.scss";
import ReportedClaimsTooltip from "../custom-components/tooltip/templates/reported-claims-tooltip";

const GroupedBarChart = props => {
  const {chart, classes, containerId, data, keys, colors, currentFilter, sortingArray} = props;
  const svgRef = useRef();
  const wrapperRef = useRef();
  const dimensions = useResizeObserver(wrapperRef);

  useEffect(() => {
    const svg = select(svgRef.current);
    const {width, height} = dimensions || wrapperRef.current.getBoundingClientRect();
    const margin = {top: 5, right: 20, bottom: 30, left: 20};
    const barPadding = .2;
    const xScale0 = scaleBand().range([margin.left, width - margin.left - margin.right]).padding(barPadding);
    const xScale1 = scaleBand();
    const yScale = scaleLinear()
      .domain([0, ])
      .range([height - margin.top - margin.bottom, 0]);
    sortingArray && data.sort((a, b) => sortingArray.indexOf(a.claimType) - sortingArray.indexOf(b.claimType));

    const tooltip = d3Tip()
      .attr('class','d3-tip')
      .offset([-10, 0])
      .style("filter","opacity(0.9)")
      .style("position", "absolute")
      .style("background", "#485465")
      .style("color","white")
      .style("font-size","11")
      .style("padding", "5 15px")
      .style("border", "1px #333 solid")
      .style("border-radius", "5px")
      .html(function (event,d) {
        return renderToString(<ReportedClaimsTooltip data={{d,colors,currentFilter}}/>);
      });

    xScale0.domain(data.map(d => d[chart.key]));
    xScale1.domain(chart && chart.group && chart.group.length > 0 && chart.group.map(groupItem => groupItem.name)).range([0, xScale0.bandwidth()]);
    yScale.domain([0, max(data,
      d => {
        let max = -1;
        Object.keys(d).forEach(key => typeof d[key] === "number" && d[key] > max && (max = d[key]));
        return max;
      })
    ]);
    const yTicks = yScale.ticks().filter(tick => Number.isInteger(tick));
    const xAxis = axisBottom(xScale0).tickSize(0).tickPadding(10);
    const yAxis = axisLeft(yScale).tickValues(yTicks).tickFormat(format("d")).tickSize(-width);
    const root = svg.selectAll(`.${chart.key}`)
      .data(data)
      .enter().append("g")
      .attr("class", chart.key)
      .attr("transform", d => `translate(${xScale0(d[chart.key])},0)`);

    root.append("circle").attr("id", `${containerId}-tipfollowscursor`)
    root.call(tooltip)
    chart && chart.group && chart.group.length > 0 && chart.group.forEach(groupItem => {
      const rootInner = root.selectAll(`.bar.${groupItem.name}`)
        .data(d => [d])
        .enter()
        .append("rect")
        .attr("class", `bar ${groupItem.name}`)
        .style("fill", d => colors[groupItem.colorMapper] ? colors[groupItem.colorMapper] : colors[d.claimType])
        .attr("x", d => xScale1(groupItem.name))
        .attr("y", height)
        .attr("width", xScale1.bandwidth())
        .attr("height", 0)
        .on("mousemove", function (event, d){
          select(this).style("opacity", 0.8)
          const target = select(`#${containerId}-tipfollowscursor`)
            .attr("cx", event.offsetX - 52)
            .attr("cy", event.offsetY - 5) // 5 pixels above the cursor
            .node();
          tooltip.show(event, d, target);
        })
        .on("mouseout", function() {
          select(this).style("opacity",1);
          tooltip.hide(this)
        });
        rootInner.transition()
        .attr("height", d => height - margin.top - margin.bottom - yScale(d[groupItem.name]))
        .attr("y", d => yScale(d[groupItem.name]) + margin.top);
    })

// Add the X Axis
    svg
      .select(".x-axis")
      .attr("transform", `translate(0, ${height - margin.bottom})`)
      .call(xAxis);

// Add the Y Axis
    svg.select(".y-axis")
      .attr("transform", `translate(${margin.left}, ${margin.top})`)
      .call(yAxis)
  }, [colors, data, dimensions, keys]);

  return (
    <div className={`c-GroupedBarChart${classes ? " " + classes : ""}`} ref={wrapperRef}>
      <svg ref={svgRef} style={{width: "100%", height: "100%"}}>
        <g className="x-axis" />
        <g className="y-axis" />
      </svg>
    </div>
  );
};

GroupedBarChart.propTypes = {
  chart: PropTypes.object,
  classes: PropTypes.string,
  colors: PropTypes.object,
  data: PropTypes.oneOfType([
    PropTypes.object,
    PropTypes.array
  ]),
  keys: PropTypes.array,
  sortingArray: PropTypes.array
};

export default memo(GroupedBarChart);
