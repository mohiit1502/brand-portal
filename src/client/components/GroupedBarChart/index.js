import GroupedBarChart from "./GroupedBarChart.jsx";

export default GroupedBarChart;
