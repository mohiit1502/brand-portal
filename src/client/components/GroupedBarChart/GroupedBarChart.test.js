/* eslint-disable filenames/match-regex, no-unused-vars, no-undef */
import React from "react";
import GroupedBarChart from "./GroupedBarChart";

describe("GroupedBarChart", () => {
   it("renders without error", () => {

   });
});
