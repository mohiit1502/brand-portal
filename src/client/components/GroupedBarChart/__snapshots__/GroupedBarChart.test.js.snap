// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grouped Bar Chart Tests renders without error 1`] = `
<Memo(GroupedBarChart)
  chart={
    Object {
      "group": Array [
        Object {
          "name": "claimsCount",
        },
        Object {
          "colorMapper": "Items",
          "name": "itemsCount",
        },
      ],
      "key": "claimType",
    }
  }
  classes="c-ChartsContainer__content__body"
  colors={
    Object {
      "Copyright": "#71C5DF",
      "Counterfeit": "#1F4F94",
      "Items": "#E4B33F",
      "Patent": "#ADDDE9",
      "Trademark": "#3483B9",
    }
  }
  containerId="widget-claims-by-type"
  currentFilter={
    Object {
      "emailId": "subhadeep.dey@walmart.com",
      "orgId": "f9bccf12-49d2-4387-b047-9670ac66f7c0",
      "role": "Super Admin",
      "widget-claims-by-brand": Object {
        "dateRange": "last30days",
      },
      "widget-claims-by-type": Object {
        "dateRange": "last30days",
      },
      "widget-claims-by-user": Object {
        "dateRange": "last30days",
      },
    }
  }
  data={
    Array [
      Object {
        "claimType": "Trademark",
        "claimsCount": 2,
        "itemsCount": 2,
      },
    ]
  }
  keys={
    Array [
      "Counterfeit",
      "Trademark",
      "Copyright",
      "Patent",
      "Items",
    ]
  }
  sortingArray={
    Array [
      "Counterfeit",
      "Trademark",
      "Copyright",
      "Patent",
    ]
  }
>
  <div
    className="c-GroupedBarChart c-ChartsContainer__content__body"
  >
    <svg
      style={
        Object {
          "height": "100%",
          "width": "100%",
        }
      }
    >
      <g
        className="x-axis"
      />
      <g
        className="y-axis"
      />
    </svg>
  </div>
</Memo(GroupedBarChart)>
`;
